### PR TITLE
Modify OAuth client id to be unique across tenant instead of server and DB script

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
### Introduction

Currently the client ID is considered to be unique across the server and as a result the same client ID cannot be used in two different tenants. However, organization bound client ID is a valid requirement and we should provide this capability. This feature requires enabling tenant qualified URLs and tenanted sessions.

When tenant qualified URLs are enabled,
- Tenant will always be available in the context or thread local. Hence the application resolving logic can be modified to retrieve the application using both the consumer key and tenant ID. 
- The proposed solution is to modify the underline DAO implementation to include the tenant ID in places where it query for apps by giving the client ID. All the DB queries will be modified to include the tenant ID.
- The tenant ID will be retrieved using a utility method added to the “IdentityTenantUtil” class.

when tenant-qualified URLs are disabled,
- The tenant won’t be available in the context or thread local.
- The client ID will be considered unique across the server and firstly the tenant has to be resolved.
- The proposed solution is to perform this at the valve level. We’ll be introducing a new valve and when a request hits that valve, if that request is an oauth request and if tenant-qualified urls are disabled, the client id will be extracted and app tenant will be queried. The valve will only be enabled for non tenant qualified cases. 

### Proposed changes in this pull request
- Introduce the tenant resolving method to IdentityTenantUtils.
- Modify the DB constraints including unique key constraints and foreign key constraints.
- Modify OAuthApplicationDAO to include tenant ID.

### Related Issue
- https://github.com/wso2/product-is/issues/16370

### Related PRs
- https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2131
- https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/240
- https://github.com/wso2/product-is/pull/16632

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/resources/dbscripts/h2.sql [hunk 1]
```diff
@@ -23,7 +23,7 @@ CREATE TABLE IF NOT EXISTS IDN_OAUTH_CONSUMER_APPS (
             APP_ACCESS_TOKEN_EXPIRE_TIME BIGINT DEFAULT 3600,
             REFRESH_TOKEN_EXPIRE_TIME BIGINT DEFAULT 84600,
             ID_TOKEN_EXPIRE_TIME BIGINT DEFAULT 3600,
-            CONSTRAINT CONSUMER_KEY_CONSTRAINT UNIQUE (CONSUMER_KEY),
+            CONSTRAINT UK_TENANT_CONSUMER_KEY UNIQUE (TENANT_ID, CONSUMER_KEY),
```
Comment (ThaminduDilshan) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/resources/dbscripts/h2.sql:26]: Use the same constraint name `CONSUMER_KEY_CONSTRAINT`

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/resources/dbscripts/h2.sql [hunk 2]
```diff
@@ -1238,7 +1238,7 @@ CREATE INDEX IDX_OCA_UM_TID_UD_APN ON IDN_OAUTH_CONSUMER_APPS(USERNAME,TENANT_ID
 CREATE INDEX IDX_SPI_APP ON SP_INBOUND_AUTH(APP_ID);
 
 -- IDN_OIDC_PROPERTY --
-CREATE INDEX IDX_IOP_CK ON IDN_OIDC_PROPERTY(CONSUMER_KEY);
+CREATE INDEX IDX_IOP_TENANT_CK ON IDN_OIDC_PROPERTY(TENANT_ID, CONSUMER_KEY);
```
Comment (ThaminduDilshan) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/resources/dbscripts/h2.sql:1241]: Use the same constraint name `IDX_IOP_CK`

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java [hunk 3]
```diff
@@ -275,7 +275,8 @@ public class ApplicationMgtDBQueries {
     public static final String REMOVE_UM_ROLE_PERMISSION = "DELETE FROM UM_ROLE_PERMISSION WHERE UM_PERMISSION_ID = ?";
 
     // DELETE query - Oauth
-    public static final String REMOVE_OAUTH_APPLICATION = "DELETE FROM IDN_OAUTH_CONSUMER_APPS WHERE CONSUMER_KEY=?";
+    public static final String REMOVE_OAUTH_APPLICATION = "DELETE FROM IDN_OAUTH_CONSUMER_APPS WHERE CONSUMER_KEY=?" +
```
Comment (ThaminduDilshan) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java:278]: - Introduce a new query and keep the existing one.
- depracate the query

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java [hunk 4]
```diff
@@ -266,6 +266,7 @@ public class ApplicationMgtDBQueries {
     public static final String REMOVE_UM_ROLE_PERMISSION = "DELETE FROM UM_ROLE_PERMISSION WHERE UM_PERMISSION_ID = ?";
 
     // DELETE query - Oauth
+    @Deprecated
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:269]: what's the new query to be used here?

Comment (ThaminduDilshan) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:269]: I don't think we can add a deprecation note as this is a variable. Added a comment

Comment (ThaminduDilshan) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:269]: Fixed

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java [hunk 5]
```diff
@@ -274,7 +274,10 @@ public class ApplicationMgtDBQueries {
     public static final String REMOVE_UM_PERMISSIONS = "DELETE FROM UM_PERMISSION WHERE UM_ID = ?";
     public static final String REMOVE_UM_ROLE_PERMISSION = "DELETE FROM UM_ROLE_PERMISSION WHERE UM_PERMISSION_ID = ?";
 
-    // DELETE query - Oauth
+    // DELETE query - OAuth.
+    @Deprecated
+    // Use the REMOVE_OAUTH_APPLICATION_WITH_TENANT query in
+    // org.wso2.carbon.identity.application.mgt.dao.impl.ApplicationMgtDBQueries class.
```
Comment (chamathns) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java:280]: let's follow standard javaDoc notation
```suggestion
    /**
 * @deprecated Use the REMOVE_OAUTH_APPLICATION_WITH_TENANT query in
 * {@link org.wso2.carbon.identity.application.mgt.dao.impl.ApplicationMgtDBQueries} class.
 */
```

**Review by chamathns — COMMENTED** [review]
Can you add a comment regarding the migration impact these schema changes would have? Just to know where we stand.

**Comment by ThaminduDilshan** [comment]
> Can you add a comment regarding the migration impact these schema changes would have? Just to know where we stand.
- For those who are migrating from older versions, there won't be a migration as long as they don't need this feature. If they need this feature, will be providing a schema migrator.
- For those who are planning to stay with the previous behavior (non tenant qualified urls and non tenanted sessions) in the new product release, need to update the DB constraints to the previous one. Can use the same migrator

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql [hunk 6]
```diff
@@ -9,7 +9,7 @@ CREATE TABLE IDN_OAUTH_CONSUMER_APPS (
             CONSUMER_KEY VARCHAR (255) NOT NULL,
             CONSUMER_SECRET VARCHAR (2048),
             USERNAME VARCHAR (255),
-            TENANT_ID INTEGER DEFAULT 0,
+            TENANT_ID INTEGER DEFAULT 0 NOT NULL,
```
Comment (dewniMW) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql:12]: is this change required?

Comment (ThaminduDilshan) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql:12]: Yes. DB2 doesn't allow to define a unique key constraint over non null columns eventhough it's defined with a default value. Tested locally and previous schema didn't work. Hence had to add `NOT NULL`

**Review by dewniMW — APPROVED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java [hunk 7]
```diff
@@ -274,7 +274,12 @@ public class ApplicationMgtDBQueries {
...
     public static final String REMOVE_UM_ROLE_PERMISSION = "DELETE FROM UM_ROLE_PERMISSION WHERE UM_PERMISSION_ID = ?";
 
-    // DELETE query - Oauth
+    // DELETE query - OAuth.
+    /**
+     * @deprecated Use the REMOVE_OAUTH_APPLICATION_WITH_TENANT query in
+     * {@link org.wso2.carbon.identity.application.mgt.dao.impl.ApplicationMgtDBQueries} class.
+     */
+    @Deprecated
     public static final String REMOVE_OAUTH_APPLICATION = "DELETE FROM IDN_OAUTH_CONSUMER_APPS WHERE CONSUMER_KEY=?";
```
Comment (pulasthi7) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java:283]: use the deprecated javadoc comment like below
    /**
     * @deprecated use {@link org.wso2.carbon.identity.application.mgt.dao.impl.ApplicationMgtDBQueries}
     */

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java [hunk 8]
```diff
@@ -266,7 +266,10 @@ public class ApplicationMgtDBQueries {
...
     public static final String REMOVE_OAUTH_APPLICATION = "DELETE FROM IDN_OAUTH_CONSUMER_APPS WHERE CONSUMER_KEY=?";
+    public static final String REMOVE_OAUTH_APPLICATION_WITH_TENANT = "DELETE FROM IDN_OAUTH_CONSUMER_APPS WHERE " +
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:271]: It doesn't specifically for tenant qualified scenarios.. Idea is to remove app with both consumer key and tenant

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java (hunk 7 above)
Comment (ThaminduDilshan) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtDBQueries.java:283]: This is already addressed
